[Update] Support Page Sections & Sphinx Domains

dojutsu-user opened the pull request:
Demo

![Imgur](https://i.imgur.com/1IMpJAU.gif)
(https://imgur.com/a/18MrgzN)

Related PR in readthedocs.org -- https://github.com/readthedocs/readthedocs.org/pull/5829

ericholscher commented on sphinx_search/_static/js/rtd_sphinx_search.js:
```diff
@@ -71,97 +75,248 @@ const createDomNode = (nodeName, attributes) => {
+            outer_div.appendChild(section_content);
+            section_link.appendChild(outer_div);
+            content.appendChild(section_link);
+            content.appendChild(createDomNode("br", { class: "br-for-hits" }));
```
I really dislike this way of creating HTML. Is it possible for us to use a JS template language for this rendering, instead of all being in code? I feel that would be much, much cleaner.

dojutsu-user replied:
I searched on the internet and found a very small and easy to use library -- [squirrelly.js](https://squirrelly.js.org/).
I have updated the PR to make use of it.
I think we can use it in other places also. So I'll be updating the code and use it where ever it proves to be effective.

davidfischer replied:
I don't have a particularly strong opinion but sphinx docs already rely on underscore which has [built-in templating](https://underscorejs.org/#template). It may not be the right fit but it could avoid an external dependency.

dojutsu-user replied:
@davidfischer 
Thanks.
I have removed the external dependency and used the uderscore.js for it.

ericholscher commented on sphinx_search/_static/js/rtd_sphinx_search.js:
```diff
@@ -71,97 +75,248 @@ const createDomNode = (nodeName, attributes) => {
+        }
+    }
+
+    if (domains !== undefined && domains !== null) {
```
Are we always showing sections before domains in this logic?

dojutsu-user replied:
Yes.
It is because we are getting the results of sections and domains separately.
sections results are sorted and domains results are sorted -- but there is no sorting present which invloves both.

ericholscher replied:
I believe ES should return the score for each object, and we're removing it in the web API. We could probably be  passing along the score to the JS, so that we could use it more smartly for sorting.

dojutsu-user replied:
Yes. Score is present in each result.
Where should the sorting be done? server side or client side?

davidfischer replied:
I think a server side sort is reasonable so that the search results on readthedocs.org and in-doc search mostly match.

dojutsu-user replied:
I have implemented a server side sort. But I'm still concerned about its efficiency as I have not implemented any algorithm.

ericholscher commented on sphinx_search/extension.py:
```diff
@@ -17,9 +18,9 @@ def copy_asset_files(app, exception):
+    for file in CUSTOM_ASSETS_FILES:
+        if file.endswith('.min.js'):
+            app.add_js_file(file)
+        if file.endswith('.min.css'):
```
For now, I'd like to use the unminified version here, so it's easier to debug.

dojutsu-user replied:
Updated the PR.

ericholscher replied:
I wonder if we should have a setting that specifies minified or non-minified files here? We can ship like it is, but I almost always prefer full versions when doing dev.

ericholscher commented on docs/custom-design.rst:
```diff
@@ -4,46 +4,98 @@ Custom Design
+                <div class="search__result__single"><div>...</div></div>
             </div>
         </div>
     </div>
```
Maybe we should delete this content until we settle on the final HTML, then we can add it back? Alternatively, we could use `literalinclude` from Sphinx to include this directly from the HTML: https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-literalinclude

dojutsu-user replied:
I have removed it for now and we can target these things things in a separate PR to imrpove the docs and other things.
I don't think that we can use `literalinclude` here because content is being dynamically generated and removed from the DOM.

ericholscher commented on sphinx_search/_static/js/rtd_sphinx_search.js:
```diff
@@ -71,97 +75,333 @@ const createDomNode = (nodeName, attributes) => {
+    }
+
+    let section_content =
+        sectionData._source.content.substring(0, 100) + " ...";
```
Do we want to substring this here? Perhaps we should do it in the API instead. I don't have a strong feeling here though -- why are we only using 100 chars?

dojutsu-user replied:
I also don't have a particular opinion here.
But I think if we preprocess this on server side, it will reduce the size of the fetched data. So preprocessing this might decrease the time to get the results from server. What are your thoughts on this?

I used the substring with `100` because I think it is just sufficient data from a page which don't have any highlighted content.
We can increase or decrease this as per our choice. I have used this in other places also.

ericholscher commented on sphinx_search/_static/js/rtd_sphinx_search.js:
```diff
@@ -71,97 +75,333 @@ const createDomNode = (nodeName, attributes) => {
+        sectionData.highlight["sections.content"].length >= 1
+    ) {
+        section_content =
+            "... " + sectionData.highlight["sections.content"][0] + " ...";
```
Is there a reason we're accessing only the first item of the list here? Should this be a constant about how many to include?

dojutsu-user replied:
We were getting only one result per section. So there has to be only one element in the array.
But I have refactored the code now and it is much better. We can show as many results from a section as we want. Here's a demo:
![Peek 2019-07-09 13-42](https://user-images.githubusercontent.com/29149191/60871220-f0f1b800-a24f-11e9-8ba8-32f72044e92a.gif)
I have used it with maximum 3 results from a section

dojutsu-user replied:
Currently I have used the approach of getting the value at index 0 directly everywhere.
I will be refactoring them also. (search.html and main site search)

ericholscher commented on sphinx_search/_static/js/rtd_sphinx_search.js:
```diff
@@ -71,97 +75,333 @@ const createDomNode = (nodeName, attributes) => {
+            domain_type_display =
+                domainData.highlight["domains.type_display"][0];
+        }
+    }
```
Can we write a function to abstract these 3 functions that do the same thing? `get_highlight_list_data('domains.display_name`)` or similar?

ericholscher commented on sphinx_search/_static/js/rtd_sphinx_search.js:
```diff
@@ -71,97 +75,333 @@ const createDomNode = (nodeName, attributes) => {
+    }
+    if (
+        typeof domain_doc_display === "string" &&
+        domain_doc_display.length > 0
```
Can we add an `_is_string` function or similar that does this?

ericholscher commented on sphinx_search/_static/js/rtd_sphinx_search.js:
```diff
@@ -71,97 +75,333 @@ const createDomNode = (nodeName, attributes) => {
+    if (domainData.highlight !== undefined && domainData.highlight !== null) {
+        if (
+            domainData.highlight["domains.name"] !== undefined &&
+            domainData.highlight["domains.name"].length >= 1
```
Can we add an `_is_list` function that does this?

ericholscher commented on sphinx_search/_static/js/rtd_sphinx_search.js:
```diff
@@ -71,97 +75,333 @@ const createDomNode = (nodeName, attributes) => {
+    let domain_subheading = "";
+    if (
+        typeof domain_display_name === "string" &&
+        domain_display_name.length >= 2 // >= 2 because many display_names have values "-"
```
Can we fix indexing to not store this data?

dojutsu-user replied:
I think we can.
I will look into this more.

dojutsu-user replied:
@ericholscher 
This is fixed now -- https://github.com/readthedocs/readthedocs.org/pull/5829/commits/f9883021c84f2295679610639346c55e0f706680

ericholscher commented on sphinx_search/_static/js/rtd_sphinx_search.js:
```diff
@@ -71,97 +75,333 @@ const createDomNode = (nodeName, attributes) => {
+ *      </a>
+ *      <br class="br-for-hits">
+ *
+ *  </div>
```
Let's keep this in the docs, we don't need it here. It takes up a lot of space.

ericholscher commented on sphinx_search/_static/js/rtd_sphinx_search.js:
```diff
@@ -71,97 +75,333 @@ const createDomNode = (nodeName, attributes) => {
+                page_link
+            );
+        } else {
+            if (type === "domains") {
```
should be `else if` ?

dojutsu-user commented:
@ericholscher 
I have updated the PR.

ericholscher commented on sphinx_search/_static/js/rtd_sphinx_search.js:
```diff
@@ -89,31 +117,41 @@ const get_section_html = (sectionData, page_link) => {
-        section_content =
-            "... " + sectionData.highlight["sections.content"][0] + " ...";
+    let section_content = [
+        sectionData._source.content.substring(0, 100) + " ..."
```
I'd like to see `100` be a constant defined at the top of the file, at least. We can keep that for now.

dojutsu-user replied:
Updated the PR.

dojutsu-user commented:
@ericholscher 
I have also updated the PR to include minified files now.

dojutsu-user commented:
@ericholscher 
This is ready to be merged I think.

ericholscher commented on MANIFEST.in:
```diff
@@ -1,6 +1,4 @@
 include LICENSE
-include sphinx_search/_static/js/rtd_sphinx_search.js
 include sphinx_search/_static/js/rtd_sphinx_search.min.js
-include sphinx_search/_static/css/rtd_sphinx_search.css
```
I do think we should include these as well as the minified ones. People generally like to be able to debug against non-minified source.

ericholscher commented on tests/test_extension.py:
```diff
@@ -36,9 +36,9 @@ def test_static_files_injected_in_html(selenium, app, status, warning):
 
     assert (
-        page_source.count('rtd_sphinx_search.js') == 1
+        page_source.count('rtd_sphinx_search.min.js') == 1
```
these should be variables, not strings put thoughout the file.

dojutsu-user commented:
@ericholscher 
I have updated the PR.
I have added a configuration variable - `RTD_SPHINX_SEARCH_FILE_TYPE` whose possible values are - `MINIFIED` and `UN_MINIFIED`. User can set these two values in their `conf.py` and the corresponding css and js files will get added in the docs.
For now, the default value is set to `UN_MINIFIED` for debugging purpose.
We will change this value once we decide to ship it.
Also, the tests for the same are added.

ericholscher commented on sphinx_search/extension.py:
```diff
@@ -1,26 +1,43 @@
+
+def setup(app):
+
+    app.add_config_value('RTD_SPHINX_SEARCH_FILE_TYPE', 'UN_MINIFIED', 'html')
```
I believe Sphinx uses lower-case config values by convention.

ericholscher replied:
This also should be documented.

dojutsu-user replied:
Can the documentation changes be done in another PR with the API reference?

ericholscher commented on sphinx_search/extension.py:
```diff
@@ -1,26 +1,43 @@
-    os.path.join('css', 'rtd_sphinx_search.min.css'),
-]
+CUSTOM_ASSETS_FILES = {
+    'MINIFIED': [
```
these should also be lowercase.

ericholscher commented on sphinx_search/extension.py:
```diff
@@ -1,26 +1,43 @@
-def setup(app):
+def inject_static_files(app):
+    """Inject correct CSS and JS files based on the value of ``RTD_SPHINX_SEARCH_FILE_TYPE``.""" 
+    file_type = app.config.RTD_SPHINX_SEARCH_FILE_TYPE
```
we should check the incoming value for being one of the two we expect, and give users a good error if not.

dojutsu-user commented:
@ericholscher 
Updated the docs.
